### PR TITLE
Prevent game from starting via checkboxes with unavailable map

### PR DESCRIPTION
**Pull request by PunkPun**
Currently the player gets kicked from the multiplayer lobby

### CODE REVIEW

**Comment by PunkPun** [comment]
rebased

**Review by AspectInteractive — DISMISSED** [review]
Looks good from here.

File: OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs [hunk 1]
```diff
@@ -259,6 +259,14 @@ static void CheckAutoStart(S server)
 				if (server.LobbyInfo.Slots.All(sl => server.LobbyInfo.ClientInSlot(sl.Key) == null))
 					return;
 
+				// Does the host have the map installed?
+				if (server.ModData.MapCache[server.Map.Uid].Status != MapStatus.Available)
+				{
+					server.LobbyInfo.Clients.ForEach((c) => c.State = Session.ClientState.NotReady);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs]: Same as #20208.

**Comment by Mailaender** [comment]
image

I can't reproduce the original issue. When I remove the map from the file system (why should I do that?) then the UI becomes inconsistent: The `Start` buttons leads to nowhere and the map search does not find the map even though it is on the resource center. `Retry Search` will find it. This pull request does not seem to solve any of that.

**Comment by PunkPun** [comment]
> I can't reproduce the original issue. 

When you click on ready instead of "starting the game" and then immediately kicking the player / everyone back to multiplayer lobby the game will simply not start. This fix doesn't allow the game to start in the first place

> When I remove the map from the file system (why should I do that?) then the UI becomes inconsistent: The `Start` buttons leads to nowhere

Start button is disabled. I'm not sure what you mean

> and the map search does not find the map even though it is on the resource center. `Retry Search` will find it. This pull request does not seem to solve any of that.

`MapPreviewLogic` needs a complete refactor. It's coded quite poorly IMO and I can't really trigger `Retry Search` without making it super hacky. For now I added a trigger to update MapCache on the `Retry Search` button. I'll have to add logic to search for last updated map with the refactor when I come to it

**Comment by Mailaender** [comment]
When I join a game, delete the map on both server and client and tick the ready checkboxes it now crashes the server:

```
/home/matthias/Entwicklung/OpenRA/mods/cnc/maps/master-alert.oramap was deleted
Exception of type `System.InvalidOperationException`: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
   at OpenRA.Mods.Common.Server.LobbyCommands.CheckAutoStart(Server server) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:line 266
   at OpenRA.Mods.Common.Server.LobbyCommands.State(Server server, Connection conn, Client client, String s) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:line 293
   at OpenRA.Mods.Common.Server.LobbyCommands.InterpretCommand(Server server, Connection conn, Client client, String cmd) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:line 236
   at OpenRA.Server.Server.<>c__DisplayClass85_0.<InterpretServerOrder>b__0(IInterpretCommand t) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 995
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
   at OpenRA.Server.Server.InterpretServerOrder(Connection conn, Order o)
   at OpenRA.Server.Server.InterpretServerOrders(Connection conn, Byte[] data) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 931
   at OpenRA.Server.Server.ReceiveOrders(Connection conn, Int32 frame, Byte[] data) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 892
   at OpenRA.Server.Server.ConnectionPacketEvent.OpenRA.Server.Server.IServerEvent.Invoke(Server server) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 1474
   at OpenRA.Server.Server.<>c__DisplayClass54_0.<.ctor>b__2(Object _) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 361
   at System.Threading.Thread.StartCallback()

Unhandled exception. System.InvalidOperationException: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
   at OpenRA.Mods.Common.Server.LobbyCommands.CheckAutoStart(Server server) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:line 266
   at OpenRA.Mods.Common.Server.LobbyCommands.State(Server server, Connection conn, Client client, String s) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:line 293
   at OpenRA.Mods.Common.Server.LobbyCommands.InterpretCommand(Server server, Connection conn, Client client, String cmd) in /home/matthias/Entwicklung/OpenRA/OpenRA.Mods.Common/ServerTraits/LobbyCommands.cs:line 236
   at OpenRA.Server.Server.<>c__DisplayClass85_0.<InterpretServerOrder>b__0(IInterpretCommand t) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 995
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
   at OpenRA.Server.Server.InterpretServerOrder(Connection conn, Order o)
   at OpenRA.Server.Server.InterpretServerOrders(Connection conn, Byte[] data) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 931
   at OpenRA.Server.Server.ReceiveOrders(Connection conn, Int32 frame, Byte[] data) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 892
   at OpenRA.Server.Server.ConnectionPacketEvent.OpenRA.Server.Server.IServerEvent.Invoke(Server server) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 1474
   at OpenRA.Server.Server.<>c__DisplayClass54_0.<.ctor>b__2(Object _) in /home/matthias/Entwicklung/OpenRA/OpenRA.Game/Server/Server.cs:line 361
   at System.Threading.Thread.StartCallback()
./launch-game.sh: Zeile 35: 225450 Abgebrochen             (Speicherabzug geschrieben) ${RUNTIME_LAUNCHER} ${ENGINEDIR}/bin/OpenRA.dll Engine.EngineDir=".." Engine.LaunchPath="${LAUNCHPATH}" ${MODARG} "$@"
```

**Comment by PunkPun** [comment]
That's a weird crash. @teinarss mentioned that host will always be client 0. But in your case client 0 didn't exist. I handled it but I may be able to handle it better if I knew why client 0 didn't exist

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d14f4d2064fe91a66d25403b416c023b190f7e4c...dd121e381475e78c9125b89b54ecc2e41d4fb076)

**Comment by anvilvapre** [comment]
Note there is a difference between `Connection.LocalClientId` is not the same as `Player.Index`.

`Game.IsHost` shows this. 

It can be confusing to see IsHost, IsAdmin, clientIndex 0 and clientAt being used to get an admin. And connection and player also having an index field(?).

**Pull request merged**